src/workerd/api/node/context.c++

```diff
@@ -0,0 +1,170 @@
+    auto context = isolate->GetCurrentContext();
+    if (!v8::Promise::Resolver::New(context).ToLocal(&resolver)) {
+      // Return empty local and allow error to propagate.
+      return v8::Local<v8::Promise>();
+    }
+    if (!resolver->Resolve(context, v8::Undefined(isolate)).IsJust()) {
+      // Return empty local and allow error to propagate.
+      return v8::Local<v8::Promise>();
+    }
+    return resolver->GetPromise();
+  };
+
+  v8::MaybeLocal<v8::Value> result;
+  v8::TryCatch catcher(isolate);
+  auto& js = jsg::Lock::from(isolate);
+  try {
+    info.evalFunc(js);
+  } catch (const JsExceptionThrown&) {
+    if (catcher.CanContinue()) catcher.ReThrow();
+    // leave `result` empty to propagate the JS exception
+    return v8::MaybeLocal<v8::Value>();
+  }
+
+  auto ctx = static_cast<api::node::NodeJsModuleContext*>(info.moduleContext.get());
+
+  if (module->SetSyntheticModuleExport(
+          isolate,
+          v8StrIntern(isolate, "default"_kj),
+          ctx->module->getExports(js.v8Isolate)).IsJust()) {
+    result = makeResolvedPromise();
+  } else {
+    // leave `result` empty to propagate the JS exception
+  }
+  return result;
+}
+
+}  // namespace workerd::jsg
+
+namespace workerd::api::node {
+
+v8::Local<v8::Value> NodeJsModuleContext::require(kj::String specifier, v8::Isolate* isolate) {
+  static const std::set<kj::StringPtr> NODEJS_BUILTINS {
+    "_http_agent",         "_http_client",        "_http_common",
+    "_http_incoming",      "_http_outgoing",      "_http_server",
+    "_stream_duplex",      "_stream_passthrough", "_stream_readable",
+    "_stream_transform",   "_stream_wrap",        "_stream_writable",
+    "_tls_common",         "_tls_wrap",           "assert",
+    "assert/strict",       "async_hooks",         "buffer",
+    "child_process",       "cluster",             "console",
+    "constants",           "crypto",              "dgram",
+    "diagnostics_channel", "dns",                 "dns/promises",
+    "domain",              "events",              "fs",
+    "fs/promises",         "http",                "http2",
+    "https",               "inspector",           "inspector/promises",
+    "module",              "net",                 "os",
+    "path",                "path/posix",          "path/win32",
+    "perf_hooks",          "process",             "punycode",
+    "querystring",         "readline",            "readline/promises",
+    "repl",                "stream",              "stream/consumers",
+    "stream/promises",     "stream/web",          "string_decoder",
+    "sys",                 "timers",              "timers/promises",
+    "tls",                 "trace_events",        "tty",
+    "url",                 "util",                "util/types",
+    "v8",                  "vm",                  "worker_threads",
+    "zlib"
+  };
+
+  // If it is a bare specifier known to be a Node.js built-in, then prefix the
+  // specifier with node:
+  bool isNodeBuiltin = false;
```

Comment:
I wonder if it would be easier to maintain to have a module like `node/internal/globals.ts` and make all exports from there a global?

Reply:
oo, I like that idea.

Reply:
this is likely to be a bit complicated given how `require` needs to work. I'll look at making this change separately.